File: 1046-last-stone-weight/1046-last-stone-weight.py
import heapq

class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:

        n=len(stones)

        pq=[]          # Max Heap Dec

        # Inserting all the Stones into the Heap
        for i in range(n):
            heapq.heappush(pq,-stones[i])

        # Jab tak Heap mein 1 se Zyada Stones hain
        while len(pq)>1:

            first=-heapq.heappop(pq)          # Largest Stone

            second=-heapq.heappop(pq)         # Second Largest Stone

            if first==second:
                # Stones are Destroyed
                continue

            # New Wt = y-x
            heapq.heappush(pq,-(first-second))

        # Agar Heap Empty hai
        if len(pq)==0:
            return 0

        return -pq[0]          # The Last Remaining Stone

File: 1046-last-stone-weight/1046-last-stone-weight.java
class Solution {
    public int lastStoneWeight(int[] stones) {

        int n=stones.length;

        PriorityQueue<Integer> pq=new PriorityQueue<>(Collections.reverseOrder());     // Max Heap Dec

        // Inserting all the Stones into the Heap
        for(int i=0;i<n;i++){
            pq.offer(stones[i]);
        }

        // Jab tak Heap mein 1 se Zyada Stones hain
        while(pq.size()>1){

            int first=pq.poll();          // Largest Stone

            int second=pq.poll();         // Second Largest Stone

            if(first==second){
                // Stones are Destroyed
                continue;
            }

            // New Wt = y-x
            pq.offer(first-second);
        }

        // Agar Heap Empty hai
        if(pq.isEmpty()){
            return 0;
        }

        return pq.peek();          // The Last Remaining Stone
    }
}